src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs

```diff
@@ -844,16 +844,95 @@ public override BoundNode VisitIsPatternExpression(BoundIsPatternExpression node
             Debug.Assert(!IsConditionalState);
             LearnFromAnyNullPatterns(node.Expression, node.Pattern);
             VisitPatternForRewriting(node.Pattern);
-            Visit(node.Expression);
-            var expressionState = ResultType;
-            var state = PossiblyConditionalState.Create(this);
-            var labelStateMap = LearnFromDecisionDag(node.Syntax, node.DecisionDag, node.Expression, expressionState, ref state);
-            var trueState = labelStateMap.TryGetValue(node.IsNegated ? node.WhenFalseLabel : node.WhenTrueLabel, out var s1) ? s1.state : UnreachableState();
-            var falseState = labelStateMap.TryGetValue(node.IsNegated ? node.WhenTrueLabel : node.WhenFalseLabel, out var s2) ? s2.state : UnreachableState();
-            labelStateMap.Free();
-            SetConditionalState(trueState, falseState);
+            if (!VisitPossibleConditionalAccess(node.Expression, out var conditionalStateWhenNotNull)
+                || !trySetStateWhenNotNull(node.Pattern, conditionalStateWhenNotNull))
+            {
+                var expressionState = ResultType;
+                var state = PossiblyConditionalState.Create(this);
+                var labelStateMap = LearnFromDecisionDag(node.Syntax, node.DecisionDag, node.Expression, expressionState, ref state);
+                var trueState = labelStateMap.TryGetValue(node.IsNegated ? node.WhenFalseLabel : node.WhenTrueLabel, out var s1) ? s1.state : UnreachableState();
+                var falseState = labelStateMap.TryGetValue(node.IsNegated ? node.WhenTrueLabel : node.WhenFalseLabel, out var s2) ? s2.state : UnreachableState();
+                labelStateMap.Free();
+                SetConditionalState(trueState, falseState);
+            }
             SetNotNullResult(node);
             return null;
+
+            bool trySetStateWhenNotNull(BoundPattern pattern, PossiblyConditionalState conditionalStateWhenNotNull)
+            {
+                Debug.Assert(!IsConditionalState);
+
+                if (conditionalStateWhenNotNull.IsConditionalState)
+                {
+                    switch (getMatchingBoolValues(pattern))
+                    {
+                        case (hasTrue: true, hasFalse: false, hasNull: false):
+                            SetConditionalState(conditionalStateWhenNotNull.StateWhenTrue, State);
+                            return true;
+                        case (hasTrue: false, hasFalse: true, hasNull: false):
+                            SetConditionalState(conditionalStateWhenNotNull.StateWhenFalse, State);
+                            return true;
+                        case (hasTrue: false, hasFalse: true, hasNull: true):
```

Comment:
Consider adding examples for the cases that aren't produced simply out of `getMatchingBoolValues`. I'm not sure how to produce this combination, for instance.

Reply:
Two ways include:
- `is not true`
- `is false or null`

I can try to comment these cases to make it more clear.

Reply:
I didn't follow for `is not true`: `true` is `(true, true, false)`, which the `not` negates into `(false, false, true)`.

For `is false or null`, what is there to learn? Seems we'd get the worst case in both branches, no?

Yes, examples would help. Thanks

Reply:
Have removed this method.